### ISSUE TITLE
🔥 Consider strong typing for API

### ISSUE THREAD

**Issue by alex-shpak** [gofiber/fiber#178]
**Is your feature request related to a problem?**
Hi!

In Go (IMO) it is considered bad practice to use `interface{}` if strongly typed API can be implemented.
I understand that idea is to have similarity with the `expressjs` library, but now developer has to visit documentation/source to identify what are proper params types for the routing function.

Also I think strongly typed API is one of things developers look after when choosing framework. It also can decrease amount of runtime checks and will show errors/mistypes during compile time.

**Describe the solution you'd like**
As an example for routing function:
```go
func (app *App) Get(
    path string, 
    handler func()(ctx *Ctx), 
    middlewares ...func()(ctx *Ctx)
) App* {
    ...
}
```

**Describe alternatives you've considered**
Leave it as it is ¯\\_(ツ)_/¯

**Comment by Fenny** [comment]
Fiber started as a personal project (coming from node) to rewrite all my web apps into Go.
I tried to stick to the same workflow as Express by using interfaces, this is indeed not the Go standard.

I do not want to break the current API by changing this, it could be possible for `v2` and I would love to hear your thoughts on this.

**Comment by alex-shpak** [comment]
I see.
Well, it is up to library maintainers to decide. I feel like having non strongly typed API can turn off some Go developers. Would be nice to have `¯\_(ツ)_/¯`.

In my case I added `fiber` to dependencies, started to write `main` function to try around and hit `app.` autocompletion, saw `args ...interface{}` and was like: damn... now I need to go find docs.

And not to mention I'm a member of strongly typed camp. 😄

**Comment by hyingreborn** [comment]
agree，Fiber made it easy for me to migrate from nodejs to Go。
The reason I chose to learn Go instead of TypeScript:
- A strong typed language can reduce errors and catch exceptions more easily
- Moderate development efficiency and adequate operational performance
- Cross-compile into binaries for easy distribution and deployment

Now, it's true that docs are often needed

**Comment by Fenny** [comment]
I have discussed this with @koddr, and we want to share our conclusion on this matter.

Most methods that use the `...interface{}` argument are used for `app.METHOD` to register routes.
We think that this is straightforward because this is the first thing you learn visiting Fiber or coming from Express.

Now there are some `*Ctx` methods that also accept `interface{}` as an argument:
`func (ctx *Ctx) Body(args ...interface{}) string`
`func (ctx *Ctx) Cookies(args ...interface{}) string`
`func (ctx *Ctx) Format(args ...interface{})`

When you managed to understand the Route handlers, you are left with `3` +/- methods where you need the documentation. So I think Fiber is still a strong typed API besides these 3 methods.
But we can improve the above `*Ctx` methods by making it strong typed for `v1.8.x`.

What are your thoughts @hyingreborn, @alex-shpak, @koddr

**Comment by alex-shpak** [comment]
I like that 👍 
I guess `Static` method would be `Static(string, ...string)`. Also for method functions something like this might be good: `Get(path string, handler func(*Ctx), handlers ...func(*Ctx)` so it requires to have at least one handler, or more.

I think having strong typing could cleanup [`router.go@register`](https://github.com/gofiber/fiber/blob/master/router.go#L138) function and make it more safe.

I cant help with some PRs if you have something specific in mind.

**Comment by Fenny** [comment]
Thanks @alex-shpak for your input, all suggestions will be addressed in v1.8.0.

PS: We went with `Get(path string, handlers ...func(*Ctx))` because we don't want to confuse people with the order of where to place middleware.

```go
app.Get("/", secretpage(), auth()) 
// Might cause unintended results since auth() will run after the secretpage

app.Get("/", auth(), secretpage())
// We think with the current arg types this will be encouraged.
```

I'm closing this issue because `v1.8.0` will be released soon, feel free to re-open for more ideas/suggestions!
